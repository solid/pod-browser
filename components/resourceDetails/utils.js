/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  AUTHENTICATED_AGENT_PREDICATE,
  AUTHENTICATED_AGENT_TYPE,
} from "../../src/models/contact/authenticated";
import {
  PUBLIC_AGENT_PREDICATE,
  PUBLIC_AGENT_TYPE,
} from "../../src/models/contact/public";

export function filterPermissionsByType(permissions, type) {
  return permissions?.filter((permission) => permission.type === type) || [];
}

export function isPublicAgentOrAuthenticatedAgentType(type) {
  return type === PUBLIC_AGENT_TYPE || type === AUTHENTICATED_AGENT_TYPE;
}

export function isPublicAgentOrAuthenticatedAgentWebId(webId) {
  return (
    webId === PUBLIC_AGENT_PREDICATE || webId === AUTHENTICATED_AGENT_PREDICATE
  );
}

export function sortByAgentName(permissions) {
  return permissions.sort((a, b) => {
    return a.profile?.name?.localeCompare(b.profile?.name);
  });
}

export function preparePermissionsDataForTable(data) {
  const publicAndAuthPermissions = data?.filter((p) =>
    isPublicAgentOrAuthenticatedAgentType(p.type)
  );
  const permissionsFilteredByType = filterPermissionsByType(data, "agent");
  const permissionsSortedByName = sortByAgentName(permissionsFilteredByType);
  return publicAndAuthPermissions.concat(permissionsSortedByName);
}
