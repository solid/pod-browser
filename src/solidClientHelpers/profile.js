/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  getSolidDataset,
  getStringNoLocale,
  getProfileAll,
  getThing,
  getUrl,
  getUrlAll,
  getSourceUrl,
  getEffectiveAccess,
} from "@inrupt/solid-client";
import { foaf, rdf, schema, space, vcard, ldp, rdfs } from "rdf-namespaces";
import { getProfileIriFromContactThing, vcardExtras } from "../addressBook";

export function displayProfileName(profile) {
  if (!profile) return null;
  if (profile.names?.length) return profile.names[0];
  return profile.webId;
}

export function getProfileFromPersonThing(profileThing) {
  return {
    avatar: getUrl(profileThing, vcard.hasPhoto),
    name:
      getStringNoLocale(profileThing, foaf.name) ||
      getStringNoLocale(profileThing, vcard.fn),
    nickname:
      getStringNoLocale(profileThing, vcard.nickname) ||
      getStringNoLocale(profileThing, foaf.nick),
    webId: getProfileIriFromContactThing(profileThing),
    types: getUrlAll(profileThing, rdf.type),
  };
}

export const TYPE_MAP = {
  [foaf.Person]: getProfileFromPersonThing,
  [schema.Person]: getProfileFromPersonThing,
};

export function getProfileFromThingError(contactThing) {
  const contact = getProfileIriFromContactThing(contactThing);
  return new Error(`Cannot handle profile for contact: ${contact}`);
}

export function getProfileFromThing(contactThing) {
  const types = getUrlAll(contactThing, rdf.type);
  const profileFn = TYPE_MAP[types.find((type) => TYPE_MAP[type])];

  if (!profileFn) {
    throw getProfileFromThingError(contactThing);
  }

  return profileFn(contactThing);
}

export function getPodConnectedToProfile(profile, location) {
  const pods = profile ? profile.pods || [] : [];
  return pods.find((pod) => location.startsWith(pod));
}

export function locationIsConnectedToProfile(profile, location) {
  return !!getPodConnectedToProfile(profile, location);
}

export function packageProfile(webId, dataset, pods, inbox) {
  const profile = getThing(dataset, webId);
  return {
    ...getProfileFromPersonThing(profile),
    webId,
    dataset,
    pods: pods || getUrlAll(profile, space.storage) || [],
    inbox: inbox || getUrl(profile, ldp.inbox),
  };
}

export async function fetchProfile(webId, fetch) {
  const profiles = await getProfileAll(webId, { fetch });
  const {
    webIdProfile,
    altProfileAll: [altProfile],
  } = profiles;

  let profileDataset = webIdProfile;
  let webIdUrl = webId;

  if (altProfile) {
    webIdUrl = getSourceUrl(altProfile);
    profileDataset = altProfile;
  }

  const pods = getUrlAll(getThing(webIdProfile, webId), space.storage);
  const inbox = getUrl(getThing(webIdProfile, webId), ldp.inbox);

  return packageProfile(webIdUrl, profileDataset, pods, inbox);
}

export async function getFullProfile(webId, session) {
  const profiles = await getProfileAll(webId);
  const profile = {
    names: [],
    avatars: [],
    types: [],
    webId,
    roles: [],
    pods: [],
    organizations: [],
    editableProfileDatasets: [],
    contactInfo: {
      phones: [],
      emails: [],
    },
  };

  const profileDataItems = [
    {
      label: "names",
      getDataItem: getStringNoLocale,
      properties: [foaf.name, vcard.fn],
    },
    {
      label: "avatars",
      getDataItem: getUrl,
      properties: [vcard.hasPhoto],
    },
    {
      label: "types",
      getDataItem: getUrl,
      properties: [rdf.type],
    },
    {
      label: "roles",
      getDataItem: getStringNoLocale,
      properties: [vcard.role],
    },
    {
      label: "organizations",
      getDataItem: getStringNoLocale,
      properties: [vcardExtras("organization-name")],
    },
    {
      label: "pods",
      getDataItem: getUrl,
      properties: [space.storage],
    },
  ];

  const contactInfoDataItems = [
    {
      label: "phones",
      properties: [vcard.hasTelephone],
    },
    {
      label: "emails",
      properties: [vcard.hasEmail],
    },
  ];

  // Note: the WebID doc is handled separately from rest of documents because of the issues with authenticated fetches and checking effective access to the resource

  // step 1: find preferences file in webId doc
  const readableProfileDocuments = [...profiles.altProfileAll];
  const webIdProfileThing = getThing(profiles.webIdProfile, webId);
  const preferencesFileUrl = getUrl(webIdProfileThing, space.preferencesFile);

  // step 2: seeAlso(s) within preferences file
  if (preferencesFileUrl && webId === session.info.webId) {
    try {
      const preferencesFile = await getSolidDataset(preferencesFileUrl, {
        fetch,
      });
      const preferencesFileThing = getThing(preferencesFile, webId);
      const seeAlsoUrls = getUrlAll(preferencesFileThing, rdfs.seeAlso);
      seeAlsoUrls.forEach(async (url) => {
        const seeAlsoDocument = await getSolidDataset(url, {
          fetch: session.fetch,
        });
        if (seeAlsoDocument) readableProfileDocuments.push(seeAlsoDocument);
      });
    } catch (e) {
      // ignore, if we cannot get this we move on
    }
  }
  // step 3: seeAlso(s) within webID doc (and isPrimaryTopicOf)
  const extendedProfilesUrls = getUrlAll(webIdProfileThing, rdfs.seeAlso);

  extendedProfilesUrls.push(getUrl(webIdProfileThing, foaf.isPrimaryTopicOf));
  const extendedProfileDocuments = await Promise.all(
    extendedProfilesUrls.map((url) => {
      return (
        url &&
        getSolidDataset(url, {
          fetch: session.fetch,
        })
      );
    })
  );
  readableProfileDocuments.push(...extendedProfileDocuments);
  // try to find profile data in WebId document first:
  profileDataItems.forEach((dataItem) => {
    dataItem.properties.forEach((property) => {
      const value = dataItem.getDataItem(webIdProfileThing, property);
      if (value) profile[dataItem.label].push(value);
    });
  });
  // try the rest of the readable documents
  readableProfileDocuments.forEach((doc) => {
    if (!doc) return;
    const webIdSubjectThing = getThing(doc, webId);
    const storageThing = getThing(doc, getSourceUrl(doc));
    const thing = webIdSubjectThing ?? storageThing;
    profileDataItems.forEach((dataItem) => {
      dataItem.properties.forEach((property) => {
        const value = dataItem.getDataItem(thing, property);
        if (value) profile[dataItem.label].push(value);
      });
    });
  });

  // find the editable profile datasets
  // check if webid is editable
  const { user: profileEditingAccess } = getEffectiveAccess(
    profiles.webIdProfile
  );
  const isProfileEditable =
    profileEditingAccess.write || profileEditingAccess.append;
  if (isProfileEditable) {
    profile.editableProfileDatasets.push(profiles.webIdProfile);
  }
  // rest of documents
  readableProfileDocuments.forEach((doc) => {
    if (!doc) return;
    const { user: profileEditingAccess } = getEffectiveAccess(doc);
    const isProfileEditable =
      profileEditingAccess.write || profileEditingAccess.append;
    if (isProfileEditable) {
      profile.editableProfileDatasets.push(doc);
    }
  });

  // find contact info
  // in webid profile
  const webIdSubjectThing = getThing(profiles.webIdProfile, webId);
  contactInfoDataItems.forEach((item) => {
    item.properties.forEach((property) => {
      const contactDetailUrls =
        webIdSubjectThing && getUrlAll(webIdSubjectThing, property);
      const contactDetailThings = contactDetailUrls?.map(
        (url) => profiles.webIdProfile && getThing(profiles.webIdProfile, url)
      );
      contactDetailThings.forEach((contactDetailItem) => {
        const contactItem = {
          type: getUrl(contactDetailItem, rdf.type),
          value: getUrl(contactDetailItem, vcard.value),
        };
        profile.contactInfo[item.label].push(contactItem);
      });
    });
  });

  // in rest of readable docs
  readableProfileDocuments.forEach((doc) => {
    if (!doc) return;
    const webIdSubjectThing = getThing(doc, webId);
    const storageThing = getThing(doc, getSourceUrl(doc));
    const thing = webIdSubjectThing ?? storageThing;
    contactInfoDataItems.forEach((item) => {
      item.properties.forEach((property) => {
        const contactDetailUrls = thing && getUrlAll(thing, property);
        const contactDetailThings = contactDetailUrls?.map(
          (url) => doc && getThing(doc, url)
        );
        contactDetailThings.forEach((contactDetailItem) => {
          const contactItem = {
            type: getUrl(contactDetailItem, rdf.type),
            value: getUrl(contactDetailItem, vcard.value),
          };
          profile.contactInfo[item.label].push(contactItem);
        });
      });
    });
  });

  return profile;
}
