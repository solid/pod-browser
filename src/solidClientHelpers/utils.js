/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* eslint-disable camelcase */

import {
  createSolidDataset,
  createThing,
  getSourceUrl,
  getThing,
  getUrlAll,
  hasResourceInfo,
  setThing,
} from "@inrupt/solid-client";
import { ldp, rdf } from "rdf-namespaces";

import { parseUrl } from "../stringHelpers";

const typeNameMap = Object.keys(ldp).reduce((acc, key) => {
  const value = ldp[key];
  return {
    ...acc,
    [value]: key,
  };
}, {});

// TODO use ldp namespace when available
export const namespace = {
  ...ldp,
  ...typeNameMap,
  mtime: "http://www.w3.org/ns/posix/stat#mtime",
  "http://www.w3.org/ns/posix/stat#mtime": "mtime",
  modified: "http://purl.org/dc/terms/modified",
  "http://purl.org/dc/terms/modified": "modified",
  size: "http://www.w3.org/ns/posix/stat#size",
  "http://www.w3.org/ns/posix/stat#size": "size",
  nickname: "http://xmlns.com/foaf/0.1/nick",
  "http://xmlns.com/foaf/0.1/nick": "nickname",
  familyName: "http://xmlns.com/foaf/0.1/familyName",
  "http://xmlns.com/foaf/0.1/familyName": "familyName",
  img: "http://xmlns.com/foaf/0.1/img",
  "http://xmlns.com/foaf/0.1/img": "img",
  name: "http://xmlns.com/foaf/0.1/name",
  "http://xmlns.com/foaf/0.1/name": "name",
  hasPhoto: "http://www.w3.org/2006/vcard/ns#hasPhoto",
  "http://www.w3.org/2006/vcard/ns#hasPhoto": "hasPhoto",
  "http://www.w3.org/2006/vcard/ns#country-name": "countryName",
  countryName: "http://www.w3.org/2006/vcard/ns#country-name",
  "http://www.w3.org/2006/vcard/ns#locality": "locality",
  locality: "http://www.w3.org/2006/vcard/ns#locality",
  "http://www.w3.org/2006/vcard/ns#postal-code": "postalCode",
  postalCode: "http://www.w3.org/2006/vcard/ns#postal-code",
  "http://www.w3.org/2006/vcard/ns#region": "region",
  region: "http://www.w3.org/2006/vcard/ns#region",
  "http://www.w3.org/2006/vcard/ns#street-address": "streetAddress",
  streetAddress: "http://www.w3.org/2006/vcard/ns#street-address",
  "http://www.w3.org/ns/pim/space#preferencesFile": "preferencesFile",
  preferencesFile: "http://www.w3.org/ns/pim/space#preferencesFile",
  "https://inrupt.com/podbrowser#preferencesFile": "podBrowserPreferencesFile",
  podBrowserPreferencesFile: "https://inrupt.com/podbrowser#preferencesFile",
  "http://www.w3.org/ns/pim/space#ConfigurationFile": "ConfigurationFile",
  ConfigurationFile: "http://www.w3.org/ns/pim/space#ConfigurationFile",
};

export function isContainerIri(iri) {
  return iri.charAt(iri.length - 1) === "/";
}

export function getIriPath(iri) {
  const { pathname } = parseUrl(iri);
  return pathname.replace(/\/?$/, "");
}

export function getTypes(dataset) {
  if (hasResourceInfo(dataset)) {
    const thing = getThing(dataset, getSourceUrl(dataset));
    return getUrlAll(thing, rdf.type);
  }
  return [];
}

export function datasetIsContainer(dataset) {
  return (
    getTypes(dataset).find((type) => type.match(/container/i)) !== undefined
  );
}

export function getTypeName(rawType) {
  if (!rawType) return "";
  return namespace[rawType] || rawType;
}

export function displayTypes(types) {
  return types?.length ? types.map((t) => getTypeName(t)) : [];
}

export function createResponder() {
  const respond = (response) => ({ response });
  const error = (e) => ({ error: e });

  return { respond, error };
}

export function chain(object, ...operations) {
  return operations.reduce((acc, transform) => {
    return transform(acc);
  }, object);
}

export async function chainPromise(object, ...operations) {
  return operations.reduce(
    async (acc, transform) => {
      // eslint-disable-next-line no-return-await
      return acc.then ? transform(await acc) : Promise.resolve(transform(acc));
    },
    object.then ? await object : Promise.resolve(object)
  );
}

export function defineThing(options, ...operations) {
  return chain(createThing(options), ...operations);
}

export function defineDataset(options, ...operations) {
  return setThing(createSolidDataset(), defineThing(options, ...operations));
}

/**
 * Returns the shared characters that two strings starts with, e.g. sharedStart("bar", "baz", "bam") will return "ba".
 */
export function sharedStart(...strings) {
  const A = strings.concat().sort();
  const a1 = A[0] || "";
  const a2 = A[A.length - 1] || "";
  const L = a1.length;
  let i = 0;
  // eslint-disable-next-line no-plusplus
  while (i < L && a1.charAt(i) === a2.charAt(i)) i++;
  return a1.substring(0, i);
}

export async function serializePromises(promiseFactories) {
  return promiseFactories.reduce(async (promise, func) => {
    const oldResult = await promise;
    const newResult = await func();
    return oldResult.concat(newResult);
  }, Promise.resolve([]));
}

export function uniqueObjects(array) {
  return Array.from(new Set(array.map(JSON.stringify))).map(JSON.parse);
}
