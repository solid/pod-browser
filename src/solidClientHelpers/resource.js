/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  createContainerAt,
  createSolidDataset,
  createThing,
  deleteFile,
  getSolidDataset,
  getSourceUrl,
  getThing,
  isContainer,
  saveSolidDatasetAt,
  setThing,
  acp_v3 as acp,
} from "@inrupt/solid-client";
// TODO: delete this when node 10 support drops"
import "core-js/proposals/promise-all-settled";
import { parseUrl } from "../stringHelpers";
import {
  getPolicyResourceUrl,
  getPolicyUrl,
  getResourcePoliciesContainerPath,
  deletePoliciesContainer,
} from "../models/policy";
import { createResponder, isContainerIri } from "./utils";
import { ERROR_CODES, isHTTPError } from "../error";
import { hasAcpConfiguration } from "../accessControl/acp/helpers/index";
import { customPolicies, namedPolicies } from "../../constants/policies";

export function getResourceName(iri) {
  let { pathname } = parseUrl(iri);
  if (isContainerIri(pathname)) {
    pathname = pathname.substring(0, pathname.length - 1);
  }
  const encodedURISegment =
    pathname.match(/(?!\/)(?:.(?!\/))+$/)?.toString() || "";
  return decodeURIComponent(encodedURISegment);
}

export async function getResource(iri, fetch) {
  const { respond, error } = createResponder();

  try {
    const dataset = await getSolidDataset(iri, { fetch });
    const resource = { dataset, iri };
    return respond(resource);
  } catch (e) {
    return error(e.message);
  }
}

export async function getProfileResource(iri, fetch) {
  try {
    const dataset = await getSolidDataset(iri, { fetch });
    const resource = { dataset, iri };
    return resource;
  } catch (e) {
    return { error: e.message, iri };
  }
}

export async function getOrCreateContainer(iri, fetch) {
  const { respond, error } = createResponder();
  const { response, error: getError } = await getResource(iri, fetch);
  if (response) return respond(response.dataset);
  if (getError && isHTTPError(getError, 404)) {
    try {
      return respond(await createContainerAt(iri, { fetch }));
    } catch (err) {
      return error(err.message);
    }
  }
  return error(getError);
}

export async function getOrCreateDatasetOld(iri, fetch) {
  const { respond, error } = createResponder();
  const { response, error: getError } = await getResource(iri, fetch);
  if (response) return respond(response.dataset);
  if (getError && isHTTPError(getError, ERROR_CODES.NOT_FOUND)) {
    try {
      return respond(
        await saveSolidDatasetAt(iri, createSolidDataset(), { fetch })
      );
    } catch (saveError) {
      return error(saveError.message);
    }
  }
  return error(getError);
}

export function getOrCreateThing(dataset, iri) {
  const existing = getThing(dataset, iri);
  if (existing) {
    return { thing: existing, dataset };
  }
  const created = createThing({ url: iri });
  const updatedDataset = setThing(dataset, created);
  return { thing: created, dataset: updatedDataset };
}

export function getBaseUrl(iri) {
  if (!iri) return iri;
  const url = new URL(iri);
  return url.origin + url.pathname;
}

export async function saveResource({ dataset, iri }, fetch) {
  const { respond, error } = createResponder();
  try {
    const baseIri = getSourceUrl(dataset) || getBaseUrl(iri);
    const response = await saveSolidDatasetAt(baseIri, dataset, { fetch });
    return respond(response);
  } catch (e) {
    return error(e.message);
  }
}

export async function deleteResource(
  resourceInfo,
  policiesContainerUrl,
  fetch
) {
  const iri = getSourceUrl(resourceInfo);
  await deleteFile(iri, {
    fetch,
  });
  const acrUrl = await acp.getLinkedAcrUrl(resourceInfo);
  const legacyAcp = !(await hasAcpConfiguration(acrUrl, fetch));
  if (!policiesContainerUrl) return;
  const policyUrl = getPolicyUrl(resourceInfo, policiesContainerUrl, legacyAcp);
  const resourcePoliciesContainerPath = getResourcePoliciesContainerPath(
    resourceInfo,
    policiesContainerUrl,
    legacyAcp
  );

  const policiesUrls = namedPolicies.concat(customPolicies).map(({ name }) => {
    return getPolicyResourceUrl(
      resourceInfo,
      policiesContainerUrl,
      name,
      legacyAcp
    );
  });

  if (!policyUrl && !policiesUrls.length) return;

  const urlsToDelete = [...policiesUrls, policyUrl].filter((url) =>
    Boolean(url)
  );

  Promise.allSettled(
    urlsToDelete.map(async (url) => {
      await deleteFile(url, { fetch });
    })
  )
    .then(() => {
      if (
        resourcePoliciesContainerPath &&
        isContainer(resourcePoliciesContainerPath)
      ) {
        deletePoliciesContainer(resourcePoliciesContainerPath, fetch);
      }
    })
    .catch((err) => {
      if (
        !isHTTPError(err.message, 404) &&
        !isHTTPError(err.message, 403) &&
        !isHTTPError(err.message, 409)
      ) {
        throw err;
      }
    });
}
