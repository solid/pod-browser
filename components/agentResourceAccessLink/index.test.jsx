/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { useRouter } from "next/router";
import { render } from "@testing-library/react";
import { addUrl, mockThingFrom, setStringNoLocale } from "@inrupt/solid-client";
import { ThingProvider } from "@inrupt/solid-ui-react";
import { foaf, rdf, vcard, schema } from "rdf-namespaces";
import AgentResourceAccessLink, { buildResourcesLink } from "./index";
import { chain } from "../../src/solidClientHelpers/utils";

jest.mock("next/router");

const mockedUseRouter = useRouter;

const alice = chain(mockThingFrom("https://example.com/alice"), (t) =>
  setStringNoLocale(t, vcard.fn, "Alice")
);
const bob = chain(
  mockThingFrom("https://example.com/bob"),
  (t) => setStringNoLocale(t, foaf.name, "Bob"),
  (t) => addUrl(t, rdf.type, schema.Person)
);
const iri = "/iri with spaces";

describe("AgentResourceAccessLink", () => {
  beforeEach(() => {
    mockedUseRouter.mockReturnValue({
      route: "/privacy",
    });
  });
  it("supports rendering name from vcard.fn", () => {
    const { asFragment } = render(
      <ThingProvider thing={alice}>
        <AgentResourceAccessLink iri={iri} />
      </ThingProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("supports rendering name from foaf.name", () => {
    const { asFragment } = render(
      <ThingProvider thing={bob}>
        <AgentResourceAccessLink />
      </ThingProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders iri if name is not found", () => {
    const { asFragment } = render(
      <ThingProvider thing={mockThingFrom("https://mockiri.com")}>
        <AgentResourceAccessLink />
      </ThingProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correct path for privacy contacts", () => {
    mockedUseRouter.mockReturnValueOnce({
      route: "/privacy",
    });
    const { asFragment } = render(
      <ThingProvider thing={bob}>
        <AgentResourceAccessLink />
      </ThingProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("buildProfileLink", () => {
  it("generates a valid path", () => {
    expect(buildResourcesLink(iri, "/privacy")).toEqual(
      `/privacy/resourceAccess/${encodeURIComponent(iri)}`
    );
  });
});
