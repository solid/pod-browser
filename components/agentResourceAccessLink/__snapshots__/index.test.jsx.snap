// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgentResourceAccessLink renders correct path for privacy contacts 1`] = `
<DocumentFragment>
  <a
    data-testid="resource-access-link"
    href="/privacy/person/https%3A%2F%2Fexample.com%2Fbob"
  >
    Bob
  </a>
</DocumentFragment>
`;

exports[`AgentResourceAccessLink renders iri if name is not found 1`] = `
<DocumentFragment>
  <a
    data-testid="resource-access-link"
    href="/privacy/app/https%3A%2F%2Fmockiri.com"
  >
    https://mockiri.com
  </a>
</DocumentFragment>
`;

exports[`AgentResourceAccessLink supports rendering name from foaf.name 1`] = `
<DocumentFragment>
  <a
    data-testid="resource-access-link"
    href="/privacy/person/https%3A%2F%2Fexample.com%2Fbob"
  >
    Bob
  </a>
</DocumentFragment>
`;

exports[`AgentResourceAccessLink supports rendering name from vcard.fn 1`] = `
<DocumentFragment>
  <a
    data-testid="resource-access-link"
    href="/privacy/app/https%3A%2F%2Fexample.com%2Falice"
  >
    Alice
  </a>
</DocumentFragment>
`;
